factor/telegram/classes/factor.php

```diff
@@ -0,0 +1,315 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Email factor class.
+ *
+ * @package     factor_telegram
+ * @subpackage  tool_mfa
+ * @author      Jan Dageförde, Laura Troost
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace factor_telegram;
+
+defined('MOODLE_INTERNAL') || die();
+
+use tool_mfa\local\factor\object_factor_base;
+
+class factor extends object_factor_base {
+    /**
+     * User input for the generated code.
+     *
+     * {@inheritDoc}
+     */
+    public function login_form_definition($mform) {
+        $mform->addElement('text', 'verificationcode', get_string('verificationcode', 'factor_telegram'));
+        $mform->setType("verificationcode", PARAM_ALPHANUM);
+        return $mform;
+    }
+
+    /**
+     * Generate a token that is sent to the user via Telegram.
+     *
+     * {@inheritDoc}
+     */
+    public function login_form_definition_after_data($mform) {
+        global $DB, $USER;
+
+        // Get the user's Telegram ID from the tool_mfa configuration.
+        $sql = 'SELECT *
+                  FROM {tool_mfa}
+                 WHERE userid = ?
+                   AND factor = \'telegram\'
+                   AND label LIKE \'telegram:%\'';
```

Comment:
This should use $DB->sql_like() for cross db likes. But is this extra where clause even needed at all?

Under what conditions would the label not match this?